File: 2353.py
import heapq
class FoodRatings:

    def __init__(self, foods: list[str], cuisines: list[str], ratings: list[int]):
        self.cuisines =cuisines
        self.size = len(foods)
        self.ranking_sys = [[] for _ in range(self.size)] 
        self.version = [0] * self.size
        self.menu = {}
        self.category = {}
        
        for i in range(self.size):
            self.menu[foods[i]] = i
            if cuisines[i] not in self.category:
                self.category[cuisines[i]] = i
            heapq.heappush(self.ranking_sys[self.category[cuisines[i]]], (-ratings[i], foods[i] ,self.version[i]))  


    def changeRating(self, food: str, newRating: int) -> None:
        self.version[self.menu[food]] += 1
        heapq.heappush(self.ranking_sys[self.category[self.cuisines[self.menu[food]]]], (-newRating, food ,self.version[self.menu[food]]))

    def highestRated(self, cuisine: str) -> str:
        while self.ranking_sys[self.category[cuisine]]:
            if self.ranking_sys[self.category[cuisine]][0][2] != self.version[self.ranking_sys[self.category[cuisine]][0][1]]: 
                heapq.heappop(self.ranking_sys[self.category[cuisine]])
            else:
                return self.ranking_sys[self.category[cuisine]][0][1]
                
        
        


# Your FoodRatings object will be instantiated and called as such:
# obj = FoodRatings(foods, cuisines, ratings)
# obj.changeRating(food,newRating)
# param_2 = obj.highestRated(cuisine)

File: Python/2353.cpp
#include <queue>
#include <string>
#include <vector>
#include <unordered_map>
#include <tuple>

using namespace std;


class FoodRatings {
    vector<priority_queue<tuple<int, string, int>, vector<tuple<int, string, int>>, greater<tuple<int, string, int>>>> rankingSys; 
    vector<int> version;
    unordered_map<string, int> catrgory;
    unordered_map<string, int> menu;

public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        int size = foods.size();
        version.resize(size, 0);
        rankingSys.resize(size, 0);
        for (int i = 0; i < size; i++)
        {
            menu[foods[i]] = i;
            if(category.find(cuisines[i]) == category.end()){
                rankingSys[i] = priority_queue<tuple<int, string, int>, vector<tuple<int, string, int>>, greater<tuple<int, string, int>>> tmp;
            }
            
        }
        
        rankingSys.
    }
    
    void changeRating(string food, int newRating) {
        
    }
    
    string highestRated(string cuisine) {
        
    }
};
